File: src/nostr/nip_44_test.rs
use super::*;

    #[test]
    fn test_nip44_padding() {
        let p1 = "short";
        let padded1 = pad(p1);
        assert_eq!(padded1.len(), 2 + 32);

        let p2 = "a".repeat(33);
        let padded2 = pad(&p2);
        assert_eq!(padded2.len(), 2 + 64);
        
        let unpadded1 = unpad(&padded1).unwrap();
        assert_eq!(unpadded1, p1);
        
        let unpadded2 = unpad(&padded2).unwrap();
        assert_eq!(unpadded2, p2);
    }

    #[test]
    fn test_nip44_roundtrip() {
        let shared_secret = [42u8; 32];
        let plaintext = "Hello NIP-44 v2!";
        let encrypted = encrypt_nip44(&shared_secret, plaintext).unwrap();
        let decrypted = decrypt_nip44(&shared_secret, &encrypted).unwrap();
        assert_eq!(plaintext, decrypted);
    }

    #[test]
    fn test_nip44_empty_message() {
        let shared_secret = [1u8; 32];
        let plaintext = "";
        let encrypted = encrypt_nip44(&shared_secret, plaintext).unwrap();
        let decrypted = decrypt_nip44(&shared_secret, &encrypted).unwrap();
        assert_eq!(plaintext, decrypted);
    }

    #[test]
    fn test_nip44_extensive_padding() {
        let shared_secret = [1u8; 32];
        // Test various lengths to trigger different power-of-two chunks
        let lengths = [1, 31, 32, 33, 127, 128, 129, 255, 256, 257, 511, 512, 513];
        for &len in &lengths {
            let plaintext = "a".repeat(len);
            let encrypted = encrypt_nip44(&shared_secret, &plaintext).unwrap();
            let decrypted = decrypt_nip44(&shared_secret, &encrypted).unwrap();
            assert_eq!(plaintext, decrypted, "Failed at length {}", len);
        }
    }

    #[test]
    fn test_nip44_mac_failure() {
        let shared_secret = vec![1u8; 32];
        let plaintext = "secret message";
        let encrypted = encrypt_nip44(&shared_secret, plaintext).unwrap();
        
        // Tamper with the ciphertext (last 32 bytes are MAC, so tamper before that)
        let mut bytes = general_purpose::STANDARD.decode(&encrypted).unwrap();
        let len = bytes.len();
        bytes[len - 33] ^= 0xff; // Flip a bit in ciphertext
        let tampered_encrypted = general_purpose::STANDARD.encode(bytes);
        
        let result = decrypt_nip44(&shared_secret, &tampered_encrypted);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: crypto failure: Invalid NIP-44 MAC");
    }

    #[test]
    fn test_nip44_unsupported_version() {
        let shared_secret = [1u8; 32];
        let mut payload = vec![0x03]; // Version 3
        payload.resize(65, 0);
        let encrypted = general_purpose::STANDARD.encode(payload);
        let result = decrypt_nip44(&shared_secret, &encrypted);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: Unsupported NIP-44 version");
    }

    #[test]
    fn test_nip44_invalid_length() {
        let shared_secret = [1u8; 32];
        let mut payload = vec![0x02]; // Too short
        payload.resize(10, 0);
        let encrypted = general_purpose::STANDARD.encode(payload);
        let result = decrypt_nip44(&shared_secret, &encrypted);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: Invalid NIP-44 payload length");
    }

    #[test]
    fn test_nip44_wrong_key() {
        let key1 = [1u8; 32];
        let key2 = [2u8; 32];
        let plaintext = "Super secret";
        let encrypted = encrypt_nip44(&key1, plaintext).unwrap();
        let result = decrypt_nip44(&key2, &encrypted);
        assert!(result.is_err());
        // Should fail MAC verification
        assert_eq!(result.unwrap_err().to_string(), "error: crypto failure: Invalid NIP-44 MAC");
    }

    #[test]
    fn test_nip44_payload_too_large() {
        let shared_secret = [1u8; 32];
        let oversized = "a".repeat(87473);
        let result = decrypt_nip44(&shared_secret, &oversized);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: NIP-44 payload too large");
    }

    #[test]
    fn test_nip44_decoded_too_large() {
        let shared_secret = [1u8; 32];
        let payload = vec![0x02u8; 65604];
        // Fill content beyond size limit
        let encrypted = base64::engine::general_purpose::STANDARD.encode(&payload);
        let result = decrypt_nip44(&shared_secret, &encrypted);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: NIP-44 decoded payload too large");
    }

    #[test]
    fn test_nip44_unpad_too_short() {
        let result = crate::nostr::nip_44::unpad(&[]);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: Invalid padding");

        let result = crate::nostr::nip_44::unpad(&[0x00]);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: Invalid padding");
    }

    #[test]
    fn test_nip44_unpad_invalid_length() {
        let padded = vec![0x00, 0x05, 0x01, 0x02]; // claims 5 bytes but only 2 available
        let result = crate::nostr::nip_44::unpad(&padded);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: Invalid padding length");
    }

File: src/nostr/nip44.test.ts
import { describe, it, expect } from 'vitest';
import { encryptNip44, decryptNip44, pad, unpad } from './nip44.ts';
import { base64Encode, base64Decode } from '../util.ts';

describe('NIP-44 padding', () => {
  it('pads short messages to 32 bytes + 2 byte header', () => {
    const padded = pad('short');
    expect(padded.length).toBe(2 + 32);
    expect(unpad(padded)).toBe('short');
  });

  it('pads 33-byte messages to next power-of-two block', () => {
    const p2 = 'a'.repeat(33);
    const padded = pad(p2);
    expect(padded.length).toBe(2 + 64);
    expect(unpad(padded)).toBe(p2);
  });
});

describe('NIP-44 roundtrip', () => {
  it('encrypts and decrypts a message', () => {
    const sharedSecret = new Uint8Array(32).fill(42);
    const encrypted = encryptNip44(sharedSecret, 'Hello NIP-44 v2!');
    const decrypted = decryptNip44(sharedSecret, encrypted);
    expect(decrypted).toBe('Hello NIP-44 v2!');
  });

  it('handles empty message', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const encrypted = encryptNip44(sharedSecret, '');
    const decrypted = decryptNip44(sharedSecret, encrypted);
    expect(decrypted).toBe('');
  });

  it('handles various message lengths', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const lengths = [1, 31, 32, 33, 127, 128, 129, 255, 256, 257, 511, 512, 513];
    for (const len of lengths) {
      const plaintext = 'a'.repeat(len);
      const encrypted = encryptNip44(sharedSecret, plaintext);
      const decrypted = decryptNip44(sharedSecret, encrypted);
      expect(decrypted).toBe(plaintext);
    }
  });
});

describe('NIP-44 error cases', () => {
  it('fails on MAC tampering', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const encrypted = encryptNip44(sharedSecret, 'secret message');
    const bytes = base64Decode(encrypted);
    bytes[bytes.length - 33] ^= 0xff;
    const tampered = base64Encode(bytes);
    expect(() => decryptNip44(sharedSecret, tampered)).toThrow('Invalid NIP-44 MAC');
  });

  it('rejects unsupported version', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const payload = new Uint8Array(65);
    payload[0] = 0x03;
    const encrypted = base64Encode(payload);
    expect(() => decryptNip44(sharedSecret, encrypted)).toThrow('Unsupported NIP-44 version');
  });

  it('rejects invalid length', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const payload = new Uint8Array(10);
    payload[0] = 0x02;
    const encrypted = base64Encode(payload);
    expect(() => decryptNip44(sharedSecret, encrypted)).toThrow('Invalid NIP-44 payload length');
  });

  it('fails with wrong key', () => {
    const key1 = new Uint8Array(32).fill(1);
    const key2 = new Uint8Array(32).fill(2);
    const encrypted = encryptNip44(key1, 'Super secret');
    expect(() => decryptNip44(key2, encrypted)).toThrow('Invalid NIP-44 MAC');
  });

  it('rejects payload too large (base64 string)', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const oversized = 'a'.repeat(87473);
    expect(() => decryptNip44(sharedSecret, oversized)).toThrow('NIP-44 payload too large');
  });

  it('rejects decoded payload too large', () => {
    const sharedSecret = new Uint8Array(32).fill(1);
    const payload = new Uint8Array(65604);
    payload[0] = 0x02;
    const encrypted = base64Encode(payload);
    expect(() => decryptNip44(sharedSecret, encrypted)).toThrow('NIP-44 decoded payload too large');
  });

  it('rejects unpad of too short data', () => {
    expect(() => unpad(new Uint8Array(0))).toThrow('Invalid padding');
    expect(() => unpad(new Uint8Array([0x00]))).toThrow('Invalid padding');
  });

  it('rejects unpad with invalid length', () => {
    const padded = new Uint8Array([0x00, 0x05, 0x01, 0x02]);
    expect(() => unpad(padded)).toThrow('Invalid padding length');
  });
});
